Prototype elasticsearch api

aroelo opened the pull request:
Closing #331

jdddog reviewed (commented):
Nice work Aniek!

Here are a few suggestions:
* Change the Python module from `elasticsearch` to `api`, as this will likely become a module with more than just the APIs used for reports / elasticsearch.
* It would be good to use CloudEndpoints for authentication because this will be more secure than us building and maintaining our own authentication solution. There is a tutorial for using CloudEndpoints with CloudRun here: https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-run. We would be using the API keys authentication method to begin with: https://cloud.google.com/endpoints/docs/openapi/authentication-method#api_keys.
* It would be good if the APIs were deployed with Terraform, so that they are deployed in the same way and along with the other observatory services.
* Why are there three different OpenAPI definitions? openapi2.yml, openapi3.yml and openapicloud2.yml

I'll give you a review of the design of the API in a future pull request, once I've had a chance to think about it more in the context of the other REST APIs that we will develop.

aroelo commented:
> Nice work Aniek!

Thanks  for the review @jdddog. Just posting my replies so they're documented on github.

> Here are a few suggestions:
> 
> * Change the Python module from `elasticsearch` to `api`, as this will likely become a module with more than just the APIs used for reports / elasticsearch.

Will do :-).

> * It would be good to use CloudEndpoints for authentication because this will be more secure than us building and maintaining our own authentication solution. There is a tutorial for using CloudEndpoints with CloudRun here: https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-run. We would be using the API keys authentication method to begin with: https://cloud.google.com/endpoints/docs/openapi/authentication-method#api_keys.

I will have a look at this starting today.

> * It would be good if the APIs were deployed with Terraform, so that they are deployed in the same way and along with the other observatory services.

I will have a look at this starting today.

> * Why are there three different OpenAPI definitions? openapi2.yml, openapi3.yml and openapicloud2.yml

I initially made the openapi3.yml, which uses openapi version 3.0.0, but saw later that this is not compatible with cloud run. This is mentioned here https://cloud.google.com/endpoints/docs/openapi as well as in the document you and Richard made regarding observatory platform infrastructure.
The openapi2.yml is more or less the same, but uses the openapi version 2.0.0 (it is more up-to-date than the initial 3.0 version).
The openapicloud2.yml was based on openapi2.yml, but used to try out cloud endpoints.

I will delete the redundant files, so there is only one openapi.yml left.

jdddog reviewed (changes requested):
Hey Aniek, it is looking good.

The main thing is that I don't think the endpoints service is working correctly yet, because you can access the API directly from the Cloud Run URL without authentication, rather than via endpoints.

> @jdddog Some issues to solve or discuss:
> **Automate Build Image**
> I don't think this can be done from Terraform. Like the compute images, perhaps this needs to be done by adding this as a CLI command, using gcloud builds (?). I have not really looked into this yet and focused on the Terraform configuration instead.

Yeah I think that we would have to do this as a CLI command. We can start by adding a dedicated command and at a later date we can make the CLI more usable by wrapping up the multiple commands / tools that are required to build and deploy the observatory into a single command.

> **Elasticsearch secrets**
> The elasticsearch username, password and addresses are obtained from google secrets. I have added them as variables in the Terraform configuration, but not yet in our observatory config-terraform.yaml file. This should be relatively straight-forward, but it's the end of the day :-)
> For now I just added the variable in Terraform Cloud manually:
> ![image](https://user-images.githubusercontent.com/22207600/106877474-c27ddc80-6713-11eb-9c36-b88101d7edc5.png)

Yeah this would be good. Do you think they should be populated as an Airflow Connection as well? It could look something like this:
```yaml
# ElasticSearch
elasticsearch:
  host: https://...
  api_key: random id
```

> **Trigger new Cloud Run revision**
> It's a bit of a pain to get a new Cloud Run revision up using Terraform, when the only thing that has changed is the docker image that sits in the google container registry. Terraform doesn't detect any changes that need to be applied.
> There is a terraform 'container_registry_image' data source (https://registry.terraform.io/providers/hashicorp/google/latest/docs/data-sources/container_registry_image), but this does not actually get into contact with the container registry:

What services need to be tainted  / re-created if the openapi specification is changed and or the Docker image is rebuilt?

 > **Unique hostname Endpoint**
> The service name of the Endpoints service needs to contain a unique suffix. So in terraform we need to create a random number id of some sort that is created/destroyed in line with the `google_endpoints_service.api` resource.
> The `google_endpoints_service.api` should be updated whenever the OpenApi config is changed. I haven't found a good way of linking all these dependencies yet.

> The error message you receive when you try to delete and then create an endpoints service with the same name:
> `Error: googleapi: Error 400: Service api-1234.endpoints.aniek-dev.cloud.goog has been deleted and will be purged after 30 days. To reuse this service, please undelete the service following https://cloud.google.com/service-infrastructure/docs/create-services#undeleting., failedPrecondition`

What would happen if we use custom domain names with Endpoints? E.g. https://cloud.google.com/endpoints/docs/openapi/serving-apis-from-domains. If we update the openapi.yml specification, can we update endpoints without destroying it and re-creating it?

> **Store configuration in separate module**
> Perhaps the resources related to the API can be stored in a separate Terraform module. They are not dependent on any of the other resources, but share ofcourse the same project id etc.

Yeah that would be great.

rhosking commented on observatory-platform/observatory/platform/api/app/api/query.py:
```diff
@@ -0,0 +1,316 @@
+http://0.0.0.0:5000/query?agg=country&subset=collaborations&api_key=158f9f48b674b0fdd18e5b465d7ae37e 268162 records,
+~80 sec
+"""
+import os
```
yes, I second this. I imagine there will soon be a fair few different use cases coming in for APIs. Particularly as we think about moving code to the oaebu repo. I could imagine there will be utiltiies, and core APIs that live in the observatory platform, some user-facing APIs that live in an OAEBU repo, and possibly some academic-observatory specific APIs that live in that repo too

aroelo replied:
I think adding the Elasticsearch related code into its own file is a good idea. Happy to rename server.py -> app.py and query.py -> elastic.py.

However, it might be good to have a separate directory for these files, away from the Dockerfile and requirements.txt.
If these are all in the same directory it means that the complete image will have to be rebuild when something changes in the Dockerfile (or any of the other files in the directory, but this is the intent), because it copies itself into the image. If we have a separate directory we can avoid this as well as having a distinction between the actual API and the files necessary to deploy the API to google cloud.

jdddog replied:
It shouldn't matter should it? If any file is changed you have to rebuild the image anyway?

aroelo replied:
Ah, yes if you build and submit it to google cloud then you will have to rebuild the image. But for local development if you just build the Dockerfile after making changes in e.g. app.py you could use the cache up until the second last line.

It's quite a small file so it does not matter too much, but I find that installing the requirements can sometimes be a bit lengthy and annoying.

rhosking reviewed (changes requested):
I think this is good, pending Jamie's comments. Like we discussed earlier, designing of the actual set of REST methods can be the next steps. Lots of interesting ideas to think about there

aroelo commented on observatory-platform/observatory/platform/api/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+# Use the official lightweight Python image.
```
Do you mean the base image or the whole Dockerfile?
For the whole Dockerfile I started with parts from the google cloud tutorial (https://cloud.google.com/run/docs/quickstarts/build-and-deploy#containerizing) , but this does not include the berglas bit, only the first 4 lines are the same.

This is the link to the base image, it doesn't seem to have a license:
https://github.com/docker-library/python/blob/afc4106edc395bed8dfe7a497444033b12452406/3.9/buster/slim/Dockerfile

It mentions this on the docker hub webpage:

> License
> View license information for Python 2 and Python 3.
> 
> As with all Docker images, these likely also contain other software which may be under other licenses (such as Bash, etc from the base distribution, along with any direct or indirect dependencies of the primary software being contained).
> 
> Some additional license information which was able to be auto-detected might be found in [the repo-info repository's python/ directory](https://github.com/docker-library/repo-info/tree/master/repos/python).
> 
> As for any pre-built image usage, it is the image user's responsibility to ensure that any use of this image complies with any relevant licenses for all software contained within.

But if you follow the link I can't find any license info there.

jdddog replied:
It should be the license from the Google Cloud tutorial Dockerfile:

```
# Copyright 2020 Google, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
```

aroelo replied:
Ah thanks. I see they left the license out in the tutorial bit, but it's included on github. That makes sense. I will add it.

aroelo commented on observatory-platform/observatory/platform/terraform/main.tf:
```diff
@@ -558,4 +558,114 @@ module "airflow_worker_vm" {
+  template {
+    spec {
+      containers {
+        image = "gcr.io/aniek-dev/api"
```
Thanks for pointing this out, will make the change

aroelo commented:
*  @jdddog You were right that the API was not set-up correctly. To fix this I added a separate Cloud Run backend deployment. In the previous set-up, the image that was used for the Cloud Run endpoints deployment was directly the Docker image that we build that contains the API. However, reading a bit more this image should actually be a  ESPv2 image including the openapi config of our endpoints deployment. 

The recommended way to make this image is by using their script which uses a base image and info from the endpoints service you just deployed (config id and cloud run hostname).

The alternative way that we're using (to avoid circular dependency) uses just the base image and an environment variable which specifies the endpoints service name, this is something like: "api-${random_id.endpoint-suffix.hex}.endpoints.${var.google_cloud.project_id}.cloud.goog"

The Cloud Run endpoints deployment now uses the x-google-backend configuration, where the backend_address is the url of the Cloud Run backend.
```
x-google-backend:
  address: ${backend_address}
  protocol: h2
```

*  Added elasticsearch to our observatory terraform configuration

*  Use elasticsearch api key instead of username/password

*  Add optional parameter to specify index date. If not specified, defaults to alias. 

*  Store terraform resources for API in separate module

*  Updated docs in readthedocs

**To do:**
> Yeah I think that we would have to do this as a CLI command. We can start by adding a dedicated command and at a later date we can make the CLI more usable by wrapping up the multiple commands / tools that are required to build and deploy the observatory into a single command.

I still need to look into this.

> What services need to be tainted / re-created if the openapi specification is changed and or the Docker image is rebuilt?

The 'google_endpoints_service.api' will be updated in place if you make a change to the openapi specification. This is working without problems at the moment. The  'google_cloud_run_service.backend' _should_ be updated when the Docker image is rebuilt. This is still a problem, as I don't know how to trigger this. I'm also not sure if it can be updated in place or needs to be destroyed/recreated.

> What would happen if we use custom domain names with Endpoints? E.g. https://cloud.google.com/endpoints/docs/openapi/serving-apis-from-domains. If we update the openapi.yml specification, can we update endpoints without destroying it and re-creating it?

I need to look into custom domain names. I don't understand how the domain name is currently working. 
In the tutorial (https://cloud.google.com/endpoints/docs/openapi/get-started-cloud-run), they use the url from Cloud Run to send requests to: `gateway-12345-uc.a.run.app`. This is also what I'm using for the current set-up. 
I have set-up the developer portal, but when I try the API there it is not working. The error is this:
ENOTFOUND: Error resolving domain "https://api-98d9421ddec0df30.endpoints.aniek-dev.cloud.goog"

Testing that the Cloud Run backend is inaccessible:
`https://api-backend-o6ogbcoyxq-uc.a.run.app/query`
![image](https://user-images.githubusercontent.com/22207600/108016922-9f75f580-704e-11eb-98e2-2d96adfd064e.png)

Testing that the Cloud Endpoints is inaccessible without API key
`https://endpoints-54d3544be0ce6cb7-o6ogbcoyxq-uw.a.run.app/query?agg=institution&subset=journals&from=2018&to=2019&journal=Molecular%20Pharmaceutics`
![image](https://user-images.githubusercontent.com/22207600/108017508-0ba52900-7050-11eb-9465-5f9acc338e28.png)

Testing that the Cloud Endpoints is accessible with API key:
`https://endpoints-54d3544be0ce6cb7-o6ogbcoyxq-uw.a.run.app/query?agg=institution&subset=journals&from=2018&to=2019&journal=Molecular%20Pharmaceutics&key=API_KEY`
![image](https://user-images.githubusercontent.com/22207600/108017363-a8b39200-704f-11eb-86a8-2e91e48890dc.png)

I have created the API key and restricted it to this API only:
![image](https://user-images.githubusercontent.com/22207600/108017627-4d35d400-7050-11eb-9e6a-431081a178aa.png)

rhosking commented on observatory-platform/observatory/platform/api/app/openapi.yml:
```diff
@@ -0,0 +1,142 @@
+#security:
+#  - APIKeyHeader: []
+#  - APIKeyQueryParam: []
+paths:
```
Just commenting here to document a conversation I just had with Jamie. While it would be good to get this first version merged without too much detailed analysis of the paths, we do need to setup ourselves up for future changes. I suggest (with exact details to be worked through, following best practice examples after some research) prefixing any method behind something roughly equivalent to url.com/version/0.1/methodName

This will allow us to maintain older versions of the API and put in place depreciation schedules, while also addressing new requirements as they come in.

rhosking commented on observatory-platform/observatory/platform/api/app/test_query.py:
```diff
@@ -0,0 +1,194 @@
+from observatory.platform.api.app.server import app
+
+
+class TestQuery(unittest.TestCase):
```
Similar to the tests, structuring the methods in such a way that enables multiple versions of the API to have tests written. Imagining a scenario in say 6 months time, where these initial tests will still be required (as we are still supporting the current version) but there may also be two other versions also running.

rhosking reviewed (changes requested):
there is also this feature of cloud run, which supports rolling updates (different to versioning of the API itself): https://cloud.google.com/run/docs/managing/revisions

Does terraform support the rolling updates? I can see this becoming useful if these APIs ever take off

aroelo commented:
> there is also this feature of cloud run, which supports rolling updates (different to versioning of the API itself): https://cloud.google.com/run/docs/managing/revisions
> 
> Does terraform support the rolling updates? I can see this becoming useful if these APIs ever take off

Yes this is possible with Terraform, we could name our revisions and assign traffic.
https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/cloud_run_service#example-usage---cloud-run-service-traffic-split

I was wondering if this would be a good way to manage versions, or use the different google projects that we have instead (dev/staging/production), or maybe a combination of both.

What I can't find from the Google revision info is if you set a percentage to split traffic, do you just get a random chance which revision you will use if you use the API, or can you manually specify this?
Not sure if I understand the purpose, but I guess you use it when you are relatively sure there are no problems with the newer revision, so you let 10% of your users use the new revision and wait to see if any problems occur? So the users are unaware which revision they use?

rhosking commented:
> I was wondering if this would be a good way to manage versions, or use the different google projects that we have instead (dev/staging/production), or maybe a combination of both.

So there a few layers to tease apart here:
- We will still need to have multiple environments (dev, staging, prod). These will likely translate into api.observatory.academy, api.staging.observatory.academy & api.dev.observatory.academy) (this is about ensuring we can develop and test code in the most similar way as possible to production, but without risking the stability of production)
- We will need to have version of the types of methods the API exposes api.observatory.academy/v/0.1/endpoints (This is about offer stability to end users)
- We will also need to use what cloud run calls versions to roll out changes in the code. (this is just something helpful to manage the risk of unintended changes, or problems that were missed in testing)

Sometimes changes to the code will just be bug fixes, so the API version doesn't change. But sometimes that are feature changes, so it might add a new method to the API, or create a new version of the API, with a different set of methods api.observatory.academy/v/0.2/endpoints

Ideally, a change is rolled out to the Cloud Run in development (as part of building out and testing the features), then once we are happy, the changes are rolled out to Cloud Run in staging, then if that goes well, it is rolled out to Cloud Run in production. The production instance in particular, is where we might want to consider the rolling updates. As you say, we should be pretty confident in the code by then, but there is always the chance end-users find a bug that we didn't anticipate, or it causes additional load on the database that we want to find out about before we move too quickly.

rhosking commented:
> What I can't find from the Google revision info is if you set a percentage to split traffic, do you just get a random chance which revision you will use if you use the API, or can you manually specify this?

Yeah, there is essentially a dice roll when each request comes in, with the chance determined by the traffic split. I believe there is a way to specifically target a version though. I think it is explained here: https://cloud.google.com/run/docs/rollouts-rollbacks-traffic-migration#tags

I need to read this all properly, but I believe you can deploy a new version, with zero traffic assigned to it, but still manually send requests using the tag

jdddog reviewed (changes requested):
Hey Aniek

Nice work, that works for me now. There are two things that would be good to address before finishing:
* Managing different versions the API
* Having predictable hostnames for the API and running develop, staging and production versions

In terms of managing API versions, it would be good to follow the recommendations that Google provides for Endpoints on it's [API lifecycle management](https://cloud.google.com/endpoints/docs/openapi/lifecycle-management) page. Most of these recommendations are more about process rather than technology.
* Different major versions of the API are deployed on different paths. E.g.
`http://api.observatory.academy/v1/query` and `http://api.observatory.academy/v2/query`.
* The minor version number is increased when backwards compatible changes are made (e.g. a new API endpoint is added) and major version is incremented when breaking changes are made (e.g. an endpoint changes how it works).
* A single backend is used to handle API requests for all major versions.

The biggest change when trying to implement this is that multiple openapi.yaml files need to be loaded at the same time into Endpoints. See step 5 on this page: https://cloud.google.com/endpoints/docs/openapi/versioning-an-api. I'm just not sure how you do that on Terraform.

To have a predictable hostnames for the API and different API environments (develop, staging and production), I think we should setup Endpoints to work with the cloud.goog domain, as described in the article [Configuring DNS on the cloud.goog domain](https://cloud.google.com/endpoints/docs/openapi/cloud-goog-dns-configure). 

 The domain names could be (it would be nice to get rid of endpoints... not sure if that is possible):
* Develop: develop.api.endpoints.PROJECT_ID.cloud.goog
* Staging: staging.api.endpoints.PROJECT_ID.cloud.goog
* Production: api.endpoints.PROJECT_ID.cloud.goog

Using the cloud.goog domain will be quicker and easier to setup than a custom domain, which is a little bit unclear from the documentation and may also require us to setup an SSL certificate ourselves; we can do this in another pull request.

jdddog commented on observatory-platform/observatory/platform/terraform/api/main.tf:
```diff
@@ -0,0 +1,163 @@
+  service_account_email = google_service_account.api-backend_service_account.email
+}
+
+resource "google_cloud_run_service" "backend" {
```
```suggestion
resource "google_cloud_run_service" "api_backend" {
```

It would be good to make the naming of this consistent with the other cloud run service.

jdddog commented on observatory-platform/observatory/platform/terraform/api/main.tf:
```diff
@@ -0,0 +1,163 @@
+}
+
+# Create/update Cloud Run service
+resource "google_cloud_run_service" "endpoints-cloud-run" {
```
```suggestion
resource "google_cloud_run_service" "api_endpoints" {
```
It would be good to make the naming of this consistent with the other cloud run service. So perhaps call this one api_endpoints and the backend api_backend?

jdddog commented on observatory-platform/observatory/platform/terraform/api/main.tf:
```diff
@@ -0,0 +1,163 @@
+
+# Create/update Cloud Run service
+resource "google_cloud_run_service" "endpoints-cloud-run" {
+  name     = "endpoints-${random_id.endpoint-suffix.hex}"
```
Does this name need to have a random id in it? Maybe calling it api-endpoints would be clearer (in line with api-backend)

jdddog reviewed (changes requested):
Hey Aniek, nice work, I've just got a few small comments below.

jdddog reviewed (changes requested):
It would be good to comment out the v2 Python code and the v2 openapi endpoint configurations so that they are just there as examples for when we need to upgrade to v2 of the API and so that people can't actually use those endpoints at the moment.

For testing the Docker image building, it would be all good to test it with code similar to: https://github.com/The-Academic-Observatory/observatory-platform/blob/develop/tests/observatory/platform/test_platform_builder.py#L284-L303. I wouldn't actually build the image with the gcloud tool, just check that it builds with Docker.
